src/Type/Php/FilterVarFunctionTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,91 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\BooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\FunctionTypeSpecifyingExtension;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\StringType;
+use function count;
+use function in_array;
+use function strtolower;
+use const FILTER_VALIDATE_BOOL;
+use const FILTER_VALIDATE_BOOLEAN;
+use const FILTER_VALIDATE_DOMAIN;
+use const FILTER_VALIDATE_EMAIL;
+use const FILTER_VALIDATE_FLOAT;
+use const FILTER_VALIDATE_INT;
+use const FILTER_VALIDATE_URL;
+
+class FilterVarFunctionTypeSpecifyingExtension implements FunctionTypeSpecifyingExtension, TypeSpecifierAwareExtension
+{
+
+	private TypeSpecifier $typeSpecifier;
+
+	public function isFunctionSupported(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		TypeSpecifierContext $context,
+	): bool
+	{
+		return strtolower($functionReflection->getName()) === 'filter_var' && $context->true();
+	}
+
+	public function specifyTypes(
+		FunctionReflection $functionReflection,
+		FuncCall $node,
+		Scope $scope,
+		TypeSpecifierContext $context,
+	): SpecifiedTypes
+	{
+		if (count($node->getArgs()) < 2) {
+			return new SpecifiedTypes();
+		}
+
+		$flagsType = $scope->getType($node->getArgs()[1]->value);
+		if (!$flagsType instanceof ConstantIntegerType) {
+			return new SpecifiedTypes();
+		}
+
+		$type = null;
+		if ($flagsType->getValue() === FILTER_VALIDATE_FLOAT) {
+			$type = new FloatType();
+		}
+		if (in_array($flagsType->getValue(), [FILTER_VALIDATE_BOOL, FILTER_VALIDATE_BOOLEAN], true)) {
+			$type = new BooleanType();
+		}
+		if (in_array($flagsType->getValue(), [FILTER_VALIDATE_DOMAIN, FILTER_VALIDATE_URL, FILTER_VALIDATE_EMAIL], true)) {
+			$type = new StringType();
+		}
```

Comment:
these might be cases that could make the impossible check type helper trigger. e.g. if the value is already a string and filter_var is used to verify that it's a domain it will specify string again. and maybe that will lead to "is always true" kind of errors. might be worth to check.
a workaround could be the FAUX function rootExpression thing again

Reply:
If I got you right, it seems the test in https://github.com/phpstan/phpstan-src/pull/1614/commits/76512faf97653c7ec08c9f256138602051fedb9b proofes it works as expected… ?

Reply:
ah sorry, I was thinking about functions that specify a boolean return type like in_array or the expressions of the webmozart/assert extension to check e.g. Assert::email. those are problematic. here it should be irrelevant, feel free to remove those tests again, I guess they don't add much value then :)

Reply:
the test I was thinking of would have been a `ImpossibleCheckTypeFunctionCallRuleTest` I think

Reply:
hmm I think this error here is related to your point:

https://github.com/composer/composer/blob/f8ebb4d8d6855457a129e01533ca3b09abb086ab/src/Composer/Util/ErrorHandler.php#L47-L50

```
 ------ --------------------------------------------------------------------- 
  Line   src/Composer/Util/ErrorHandler.php                                   
 ------ --------------------------------------------------------------------- 
  47     Call to function filter_var() with string|false and 258 will always  
         evaluate to true.                                                    
 ------ --------------------------------------------------------------------- 
 ```

Reply:
should be fixed with https://github.com/phpstan/phpstan-src/pull/1614/commits/d6937a3bf31dba83c631b8411efa4b91480084f8

Reply:
Also check maybe what happens if you mix different filters, eg first email and then domain or so. It could be that they influence each other still and it'd be better to use a faux function with the flag in the name just to be safe

Reply:
since I am passing in all `$args` to the FAUX function, which also includes the filter constant, this is fine I guess.

Reply:
True, I'm too influenced by webmozart assert internals which are different :)